### DISCUSSION TITLE
ssh and ssh-agent

### DISCUSSION THREAD

**Discussion by baoke344** [i3/i3#5495] [Q&A] [answered]
Hi! I just started using i3wm and kitty as terminal

ssh-agent is not auto start when I first login in i3 and all the previous add keys also gone. Every time I have to start ssh-agent and add all the keys again manually. Is there any way I can start ssh-agent automatically when login in to and using the cache key ( auto re-add the key) with i3

**Comment by orestisfl** [comment, marked as answer]
How do you start i3? If using `startx`, then you can start the ssh-agent in `~/.xinitrc`. If using some other session manager, you need to configure it accordingly. I think ssh-agent doesn't depend on X11 at all, so you could also start it with your login shell and export the necessary environmental variables there. What I personally do, is use `gnome-keyring-daemon` which also wraps `ssh-agent`: In my `~/.xinitrc`:

```shell
eval "$(/usr/bin/gnome-keyring-daemon --start --components=pkcs11,secrets,ssh)"
export SSH_AUTH_SOCK
```

Reference: https://wiki.archlinux.org/title/SSH_keys#SSH_agents

**Reply by baoke344** [reply]
I am using i3 on fresh install Ubuntu, so I quite sure I am on X11. Tried your method as gnome-keyring already installed
1. As ~/.xinitrc not existed, so I created 1
2. Add the lines of code into ~/.xinitrc
Still ssh-agent not started, when I tried to connect to ssh-agent it still say `Could not open a connection to your authentication agent.`

**Reply by orestisfl** [reply]
Does the output of `pidof gnome-keyring-daemon` return anything? If not, the daemon is not running.

The answer really does depend on the login manager you are using. The fact that ~/.xinitrc did not exist means that you are not using `startx` to start i3. A file that is used by many login managers is ~/.xsession but it's not a guarantee that that will work either.

**Reply by baoke344** [reply]
> Does the output of pidof gnome-keyring-daemon return anything? 

Yes it did output the pid of daemon. When I tried to run the command alone, it succeed start the ssh-agent with other error `discover_other_daemon: 1%`

So now I just need to find where to put the script and dealing with the new error

**Reply by orestisfl** [reply]
Is the SSH_AUTH_SOCK environmental variable set?

**Reply by baoke344** [reply]
Nope echo SSH_AUTH_SOCK show nothing

**Reply by baoke344** [reply]
Thanks @orestisfl for giving me so much ideal. I have came up with a work around way and it working
